src/core/vector/qgsvectorlayereditutils.cpp

```diff
@@ -735,6 +735,119 @@ int QgsVectorLayerEditUtils::addTopologicalPoints( const QgsPointXY &p )
   return addTopologicalPoints( QgsPoint( p ) );
 }
 
+bool QgsVectorLayerEditUtils::mergeFeatures( const QgsFeatureIds &mergeFeatureIds, const QgsAttributes &mergeAttributes, QgsGeometry unionGeometry, QString &errorMessage )
+{
+  errorMessage.clear();
+
+  if ( mergeFeatureIds.size() < 2 )
+  {
+    errorMessage = QObject::tr( "Not enough features to merge, the merge tool requires at least two features" );
+    return false;
+  }
+
+  QgsAttributeMap newAttributes;
+  QgsFeatureId mergeFeatureId = FID_NULL;
+  for ( int i = 0; i < mergeAttributes.count(); ++i )
+  {
+    QVariant val = mergeAttributes.at( i );
+
+    bool isDefaultValue = mLayer->fields().fieldOrigin( i ) == QgsFields::OriginProvider &&
+                          mLayer->dataProvider() &&
+                          mLayer->dataProvider()->defaultValueClause( mLayer->fields().fieldOriginIndex( i ) ) == val;
```

Comment:
@nyalldawson do we need to take care of `QgsUnsetValue` here?

Reply:
This question might be obsolete with the comment below (but still interesting nonetheless).